### ISSUE TITLE
[Feature] Add ONNX export support

### ISSUE THREAD

**Issue by iamthebot** [dmlc/xgboost#2691]
Facebook and Microsoft recently introduced [ONNX](https://github.com/onnx/onnx)-- a common IR that defines a static computational graph which can be imported in a variety of ML runtimes (like PyTorch, Caffe2, CNTK). This would let us perform inference on the runtime of our choice (eg; Caffe2) and leverage improvements in hardware and optimizations in graph compilers. Or instantly use something like Tensorflow serving (once ONNX integration is finished) to serve an XGBoost model.

Since we end up with a tree, this shouldn't be too bad to convert to an ONNX graph. At Airbnb we use XGBoost a lot and it would be nice to have it run on a single shared runtime. I'm happy to work on this but curious if there are any other takers.

**Comment by tqchen** [comment]
computational graph is quite different  from tree based models and the optimizations and representations differ a lot. The detailed runtime strategy is not likely being easily presented as a tensor op.  So in short term it might not make sense to do so

**Comment by tqchen** [comment]
To clarify what I mean, I think it is possible to computational graph to expression trees as Boolean expression, however they are less optimized as the original purpose of the tensor programming. A specific format and optimizer for tree can makes things faster for this 

This being said, there are some effort in common runtime for trees in dmlc, which should be of interest on this direction.
